web/tls_config.go

```diff
@@ -185,38 +193,25 @@ func ConfigToTLSConfig(c *TLSConfig) (*tls.Config, error) {
 		cfg.CurvePreferences = cp
 	}
 
-	if c.ClientCAs != "" {
-		clientCAPool := x509.NewCertPool()
-		clientCAFile, err := os.ReadFile(c.ClientCAs)
-		if err != nil {
-			return nil, err
-		}
-		clientCAPool.AppendCertsFromPEM(clientCAFile)
-		cfg.ClientCAs = clientCAPool
+	clientCAs, err := GetClientCAs(c.ClientCAs)
+	if err != nil {
+		return nil, err
 	}
+	cfg.ClientCAs = clientCAs
 
-	if c.ClientAllowedSans != nil {
-		// verify that the client cert contains an allowed SAN
-		cfg.VerifyPeerCertificate = c.VerifyPeerCertificate
+	clientAuth, err := ParseClientAuth(c.ClientAuth)
+	if err != nil {
+		return nil, err
 	}
 
-	switch c.ClientAuth {
-	case "RequestClientCert":
-		cfg.ClientAuth = tls.RequestClientCert
-	case "RequireAnyClientCert", "RequireClientCert": // Preserved for backwards compatibility.
-		cfg.ClientAuth = tls.RequireAnyClientCert
-	case "VerifyClientCertIfGiven":
-		cfg.ClientAuth = tls.VerifyClientCertIfGiven
-	case "RequireAndVerifyClientCert":
-		cfg.ClientAuth = tls.RequireAndVerifyClientCert
-	case "", "NoClientCert":
-		cfg.ClientAuth = tls.NoClientCert
-	default:
-		return nil, errors.New("Invalid ClientAuth: " + c.ClientAuth)
+	if c.ClientCAs != "" && clientAuth == tls.NoClientCert {
+		return nil, errors.New("Client CA's have been configured without a Client Auth Policy")
 	}
 
-	if c.ClientCAs != "" && cfg.ClientAuth == tls.NoClientCert {
-		return nil, errors.New("Client CA's have been configured without a Client Auth Policy")
+	switch clientAuth {
+	case tls.RequireAnyClientCert, tls.VerifyClientCertIfGiven, tls.RequireAndVerifyClientCert:
+		// Cert verification is delegated to the authentication middleware.
+		cfg.ClientAuth = tls.RequestClientCert
```

Comment:
can you do this unconditionally?

Reply:
There's [NoClientCert](https://cs.opensource.google/go/go/+/master:src/crypto/tls/common.go;l=301), in which case the certificate shouldn't be requested
```
        // NoClientCert indicates that no client certificate should be requested
	// during the handshake, and if any certificates are sent they will not
	// be verified.
	NoClientCert = iota
```

Reply:
I suppose this switch also depends on `ClientCAs!=""` so in that case this would be ruled out.

I general if you get a client CA you need to request the client cert, if you don't have the client CA there is no point in requesting that.